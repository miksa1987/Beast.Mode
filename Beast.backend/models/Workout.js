const mongoose = require('mongoose')

const workoutSchema = mongoose.Schema({
  description: String,
  exercises: [ { name: String, sets: Number, reps: Number, rounds: Number } ],
  picture: String,
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  likes: Number,
  comments: [ {user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' }, content: String} ]
})

module.exports = mongoose.model('Workout', workoutSchema)