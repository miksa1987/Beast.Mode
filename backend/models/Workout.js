const mongoose = require('mongoose')

const workoutSchema = mongoose.Schema({
  content: { type: String, required: true },
  picture: {Â type: String, default: '' },
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  likes: [ { type: mongoose.Schema.Types.ObjectId, ref: 'User' } ],
  likesLength: { type: Number, default: 0 },
  type: { type: String, default: 'workout' },
  comments: [ { 
    user: String,
    userid: mongoose.Schema.Types.ObjectId,
    content: String,
    date: { type: Date, default: new Date() }
  } ],
  date: { type: Date, default: new Date() },
})

module.exports = mongoose.model('Workout', workoutSchema)