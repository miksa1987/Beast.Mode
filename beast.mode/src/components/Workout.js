import React from 'react'
import { Button, Image, Divider } from 'semantic-ui-react'

const Workout = (props) => {
  if (!props.workout.content) {
    return ( <div>Loading</div> )
  }

  return ( <div>
    <table>
      <tbody>
        <tr>
          <td>
            <Image size='big' src={props.workout.picture && props.workout.picture !== '' 
              ? props.workout.picture : 'https://react.semantic-ui.com/images/wireframe/image.png'} />
          </td>
          <td>

          </td>
          <td>
            <Button onClick={() => props.history.push(`/doworkout/${props.post._id}`)}>
              Do this workout
            </Button>
          </td>
        </tr>
      </tbody>
    </table>
    
  </div> )
}

export default Workout