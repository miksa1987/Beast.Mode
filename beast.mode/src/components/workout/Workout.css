.workout-component {
  width: 98%;
  background-color: #ffffff;
  margin: 15px 15px 15px 15px;
  padding: 0px 0px 0px 0px;
  border: 1px solid #dddddd;
  border-radius: 4px;
}

.workout-image {
  width: 25%;
}

.workout-info {
  width: 75%;
  vertical-align: top;
}

.workout-button {
  width: 100%;
}

.workout-segment {
  line-break: strict;
  white-space: pre-line;
  vertical-align: top;
  padding: 10px 10px 10px 10px;
}