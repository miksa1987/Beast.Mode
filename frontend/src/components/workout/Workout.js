import React from 'react'
import { withRouter, Link } from 'react-router-dom'
import { Button, Image } from 'semantic-ui-react'
import styled from 'styled-components'
import parser from '../../service/parser'
import Animation from '../Animation'

const Base = styled.div`
  display: grid;
  grid-template: 2rem 9.3rem 2rem / 100%;
  background-color: #ffffff;
  margin: 10px 5px;
  border: 1px solid #dddddd;
`

const Columns = styled.div`
  display: grid;
  padding: 10px;
  grid-template: 100% / 5rem 5rem 5rem 1fr;
  grid-gap: 5px;
`

const ViewWorkoutLink = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.15);
`

const Workout = (props) => {
  const getExercises = () => {
    const lines = props.workout.content.split('\n')
    let exercises = []
    
    lines.forEach(line => {
      if (parser.isWorkout(line)) exercises = exercises.concat(line)
    })

    const linesToReturn = exercises.length > 8 ?
      exercises.slice(0, 5) : exercises.slice(0, exercises.length/2+1)
      
    return [linesToReturn, exercises.length]
  }

  const [exercises, exercisesNumber] = getExercises()

  if (!props.workout.content) {
    return ( <div>Loading</div> )
  }

  return (
    <Animation>
      <Base data-testid='workout-component'>
        <h3>WORKOUT BY {props.workout.user?.username?.toUpperCase()}</h3>

        <Columns>
          <Image size='tiny' rounded src={props.workout.picture && props.workout.picture !== '' 
            ? props.workout.picture : '/img/workout.jpg'} />

          <div>
            <h2>{exercisesNumber}</h2>
            exercises
          </div>

          <div>
            <h2>{props.workout.likes.length}</h2>
            likes
          </div>
          
          <div>
            {exercises.map((exercise, i) => <p key={i}>{exercise}</p>)}
          </div>
        </Columns>

        <ViewWorkoutLink>
          <Link to={`/doworkout/${props.workout._id}`}>
            <strong>View this workout</strong>
          </Link>
        </ViewWorkoutLink>
        
      </Base>
    </Animation>
  )
}

export default withRouter(Workout)