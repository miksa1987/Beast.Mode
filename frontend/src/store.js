import { createStore, combineReducers, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

import currentUserReducer from './reducers/currentUser'
import feedReducer from './reducers/feedReducer'
import workoutsReducer from './reducers/workoutsReducer'
import usersReducer from './reducers/usersReducer'
import doneWorkoutsReducer from './reducers/doneWorkoutsReducer'
import currentUserPostsReducer from './reducers/currentUserPosts'
import currentWorkoutReducer from './reducers/currentWorkout'
import currentProfileReducer from './reducers/currentProfile'
import currentPostReducer from './reducers/currentPost'
import notificationReducer from './reducers/notificationReducer'
import searchResultsReducer from './reducers/searchResultsReducer'

const reducer = combineReducers({
  currentUser: currentUserReducer,
  currentUserPosts: currentUserPostsReducer,
  currentWorkout: currentWorkoutReducer,
  currentProfile: currentProfileReducer,
  currentPost: currentPostReducer,
  users: usersReducer,
  feed: feedReducer,
  workouts: workoutsReducer,
  doneWorkouts: doneWorkoutsReducer,
  notification: notificationReducer,
  search: searchResultsReducer
})

const store = createStore(reducer, applyMiddleware(thunk))

export default store