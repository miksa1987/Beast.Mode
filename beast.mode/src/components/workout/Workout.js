import React from 'react'
import { withRouter } from 'react-router-dom'
import { Button, Image } from 'semantic-ui-react'
import parser from '../../service/parser'
import useWindowSize from '../../hooks/useWindowSize'
import './Workout.css'
import '../Animation.css'

const Workout = (props) => {
  const screen = useWindowSize()

  // Get first lines from content
  const getExercises = () => {
    const lines = props.workout.content.split('\n')
    let exercises = []
    
    lines.forEach(line => {
      if (parser.isWorkout(line)) exercises.push(line)
    })

    const linesToReturn = lines.length > 8 ?
      lines.slice(0, 5) : lines.slice(0, lines.length/2+1)
      
    return [linesToReturn, lines.length]
  }

  const [exercises, exercisesNumber] = getExercises()

  if (!props.workout.content) {
    return ( <div>Loading</div> )
  }

  return ( <div data-testid='workout-component' className='workout-component fade-in-fast'>
    <table>
      <tbody>
        <tr>
          {screen.width > screen.height && <td className='workout-image'>
            <Image width='130px' height='130px' src={props.workout.picture && props.workout.picture !== '' 
              ? props.workout.picture : 'https://react.semantic-ui.com/images/wireframe/image.png'} />
          </td>}

          <td className='workout-info'>
            <table className='third'>
              <tbody>
                <tr>
                  <td data-testid='workout-title'>
                    <h3>WORKOUT BY {props.workout.user.username.toUpperCase()}</h3>
                  </td>
                </tr>
                <tr>
                  <td>
                    <table className='third'>
                      <tbody>
                        <tr>
                          <td data-testid='workout-exnumber' className='workout-segment first'>
                            <h2>{exercisesNumber}</h2>
                            exercises
                          </td>
                          <td data-testid='workout-likes' className='workout-segment second'>
                            <h2>{props.workout.likes.length}</h2>
                            likes
                          </td>
                          <td data-testid='workout-exercises' className='workout-segment third'>
                            <ul>
                              {exercises.map((exercise, i) => <li key={i}>{exercise}</li>)}
                            </ul>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
          <td className='workout-button'>
            <Button data-testid='viewworkoutbutton' id='viewworkoutbutton' 
              onClick={() => props.history.push(`/doworkout/${props.workout._id}`)}>
              View
            </Button>
          </td>
        </tr>
      </tbody>
    </table>
  </div> )
}

export default withRouter(Workout)