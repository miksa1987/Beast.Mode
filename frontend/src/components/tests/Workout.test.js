import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import '@testing-library/jest-dom/extend-expect'
import { render, cleanup, fireEvent } from '@testing-library/react'
import Workout from '../workout/Workout'
import store from '../../store'

afterEach(cleanup)

const workout = {
  content: '5x5 pullup\n5x5 push up\n5x5 squat\n5x5 toes to bar',
  picture: '',
  comments: [],
  likes: [ '423423452', '214124124' ],
  user: {
    id: '423524242',
    username: 'Miksa'
  }
}

test('Workout component renders everything', () => {
  const component = render(
    <Provider store={store}>
      <Router>
        <Workout workout={workout}Â />
      </Router>
    </Provider>
  )
  
  expect(component).toMatchSnapshot()
})