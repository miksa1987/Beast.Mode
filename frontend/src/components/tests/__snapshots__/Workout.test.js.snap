// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout component renders everything 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdVaJa fTqoCG"
      >
        <div
          class="sc-bwzfXH lATRP"
          data-testid="workout-component"
        >
          <h3>
            WORKOUT BY 
            MIKSA
          </h3>
          <div
            class="sc-htpNat krjxIW"
          >
            <img
              class="ui tiny rounded image"
              src="/img/workout.jpg"
            />
            <div>
              <h2>
                4
              </h2>
              exercises
            </div>
            <div>
              <h2>
                2
              </h2>
              likes
            </div>
            <div>
              <p>
                5x5 pullup
              </p>
              <p>
                5x5 push up
              </p>
              <p>
                5x5 squat
              </p>
            </div>
          </div>
          <div
            class="sc-bxivhb eRfAzJ"
          >
            <a
              href="/doworkout/undefined"
            >
              <strong>
                View this workout
              </strong>
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdVaJa fTqoCG"
    >
      <div
        class="sc-bwzfXH lATRP"
        data-testid="workout-component"
      >
        <h3>
          WORKOUT BY 
          MIKSA
        </h3>
        <div
          class="sc-htpNat krjxIW"
        >
          <img
            class="ui tiny rounded image"
            src="/img/workout.jpg"
          />
          <div>
            <h2>
              4
            </h2>
            exercises
          </div>
          <div>
            <h2>
              2
            </h2>
            likes
          </div>
          <div>
            <p>
              5x5 pullup
            </p>
            <p>
              5x5 push up
            </p>
            <p>
              5x5 squat
            </p>
          </div>
        </div>
        <div
          class="sc-bxivhb eRfAzJ"
        >
          <a
            href="/doworkout/undefined"
          >
            <strong>
              View this workout
            </strong>
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
